lib/analysis/modules/ies_enkf.c

```diff
@@ -0,0 +1,803 @@
+         matrix_free( Et );
+         matrix_free( Cee );
+      } else if (ies_inversion == IES_INVERSION_SUBSPACE_EXACT_R){
+         fprintf(log_fp,"Subspace inversion using 'exact' full R. (ies_inversion=%d)\n",ies_inversion);
+         matrix_scale(R,nsc*nsc); // since enkf_linalg_lowrankCinv solves (SS' + (N-1) R)^{-1}
+         if (dbg) matrix_pretty_fprint_submat(R,"R","%11.5f",log_fp,0,m_nrobs,0,m_nrobs) ;
+         enkf_linalg_lowrankCinv( S , R , X1 , eig , truncation , subspace_dimension);
+      }
+
+      nrsing=0;
+      fprintf(log_fp,"\nEig:\n");
+      for (int i=0;i<nrmin;i++){
+         fprintf(log_fp," %f ", eig[i]);
+         if ((i+1)%20 == 0) fprintf(log_fp,"\n") ;
+         if (eig[i] < 1.0) nrsing+=1;
+      }
+      fprintf(log_fp,"\n");
+
+/*    X3 = X1 * diag(eig) * X1' * H (Similar to Eq. 14.31, Evensen (2007))                                  */
+      enkf_linalg_genX3(X3 , X1 , H , eig);
+
+      if (dbg) matrix_pretty_fprint_submat(X1,"X1","%11.5f",log_fp,0,m_nrobs,0,util_int_min(m_nrobs,nrmin-1)) ;
+      if (dbg) matrix_pretty_fprint_submat(X3,"X3","%11.5f",log_fp,0,m_nrobs,0,m_ens_size) ;
+
+/*    Update data->W = (1-ies_steplength) * data->W +  ies_steplength * S' * X3                          (Line 9)    */
+      matrix_dgemm(W0 , S , X3 , true , false , ies_steplength , 1.0-ies_steplength);
+
+      matrix_free( X1 );
+      matrix_free( X3 );
+
+   } else if (ies_inversion == IES_INVERSION_EXACT) {
+      fprintf(log_fp,"Exact inversion using diagonal R=I. (ies_inversion=%d)\n",ies_inversion);
+      matrix_type * Z      = matrix_alloc( ens_size , ens_size  );  // Eigen vectors of S'S+I
+      matrix_type * StH    = matrix_alloc( ens_size , ens_size );
+      matrix_type * StS    = matrix_alloc( ens_size , ens_size );
+      matrix_type * ZtStH  = matrix_alloc( ens_size , ens_size );
+
+      matrix_diag_set_scalar(StS,1.0);
+      matrix_dgemm(StS,S,S,true,false,1.0,1.0);
+      matrix_dgesvd(DGESVD_ALL , DGESVD_NONE , StS , eig , Z , NULL);
+
+      matrix_dgemm(StH,S,H,true,false,1.0,0.0);
+      matrix_dgemm(ZtStH,Z,StH,true,false,1.0,0.0);
+
+      for (int i=0;i<ens_size;i++){
+         eig[i]=1.0/eig[i] ;
+         matrix_scale_row(ZtStH,i,eig[i]);
+      }
+
+      fprintf(log_fp,"\nEig:\n");
+      for (int i=0;i<ens_size;i++){
+         fprintf(log_fp," %f ", eig[i]);
+         if ((i+1)%20 == 0) fprintf(log_fp,"\n") ;
+      }
+      fprintf(log_fp,"\n");
+
+/*    Update data->W = (1-ies_steplength) * data->W +  ies_steplength * Z * (Lamda^{-1}) Z' S' H         (Line 9)    */
+      matrix_dgemm(W0 , Z , ZtStH , false , false , ies_steplength , 1.0-ies_steplength);
+
+      matrix_free(Z);
+      matrix_free(StH);
+      matrix_free(StS);
+      matrix_free(ZtStH);
+   }
+
+   if (dbg) matrix_pretty_fprint_submat(W0,"Updated W","%11.5f",log_fp,0,m_ens_size,0,m_ens_size) ;
+
+
+
+/* Store active realizations from W0 to data->W */
+   {
+      int i=-1;
+      int j;
```

Comment:
@geirev : I think maybe there is a bug in the code below here; when looking at the code it seems like `dataW` should have dimensions `[total_ens_size, total_ens_size]` - but when [allocating `data->W`](https://github.com/equinor/libres/pull/422/files#diff-198bf63d37b336e1e6e5ed266b7ad034R173) the size used is `ens_size` which amount to the currently *active* ensemble size?

Reply:
Yes this sounds correct. I never got to test with failed realizations when this bug would hit hard.

Reply:
Good - then I understand.